### PR TITLE
Support for reaching top of the element

### LINKED ISSUE #2: Support for reaching top of the element
Follow up on discussion here:
https://github.com/pawel-piskorz/jquery.bum-smack/commit/a7eff5dc86780ce4681509f3445ab993ef118349

**Summary:**
_What?_
It would be nice to have smack detect when user reaches top of the element too.

_How?_
Optional attribute "edge" with values "top"/"bottom" ("bottom is default value").

_Drawbacks_
This will keep backward compatibility, but requires to change of smack scroll event namespace.

### PR DESCRIPTION
**Pull request by pawel-piskorz**
- added "edge" parameter support (default to "bottom")

Fixes #2

### CODE REVIEW

**Comment by pawel-piskorz** [comment]
Sorry for the mess, I didn't know how to add pull request to existing issue.

**Comment by mrjoelkemp** [comment]
Thanks man! Good stuff! Will look it over tonight. :)

**Comment by mrjoelkemp** [comment]
Made some commits to this pull request locally. Trying to figure out how to merge in your most recent change and push my changes up.

As an FYI, I added tests to your changes. So far, the PR fails on two tests: 
- `smacks either edge if "either" is supplied as the edge`
- `detects a pixel distance smack from the top`

I think the second one introduces a bug in the design where if you're at the top of the div, bind to a percentage threshold from the top, and then scroll down, you'll incorrectly trigger the smack.

**Comment by mrjoelkemp** [comment]
@pawel-piskorz I think the easiest is for you to give me collaborator access to your fork. I checked out your issue2 branch and merged my changes onto that branch; however, I can't push up to that branch (subsequently updating the PR) without collaborator permissions. We can then work from there to get the new features working well.

**Comment by pawel-piskorz** [comment]
@mrjoelkemp I added you as collaborator, feel free to push your changes.
As for first PR tests fails:
- I totally forgot about "either" option, right now the code supports only "top" or "bottom" for the edge param. I'll add it after you push your changes, ok?
- you're totally right, any ideas how to solve it?

BTW I had problems with installing new tests, I'll try to resolve them and run tests next time.

**Comment by mrjoelkemp** [comment]
Thanks! Away from my machine for the next couple of hours. Will push the new tests when I get back.

**Comment by mrjoelkemp** [comment]
Pushed! My stuff shows up as a [commit on the PR](https://github.com/mrjoelkemp/jquery.bum-smack/pull/5#commits-pushed-e21aa8c).

> I totally forgot about "either" option, right now the code supports only "top" or "bottom" for the edge param. I'll add it after you push your changes, ok?

No worries. 

I was a bit confused by https://github.com/mrjoelkemp/jquery.bum-smack/pull/5/files#diff-419950f0fca14a2b5fc2c6770101e81bR55 as that kills the smack binding if the edge is not 'top' or 'bottom'. It might be better to do an explicit error throw if edge is not 'top', 'bottom', or 'either' (case insensitive).

> you're totally right, any ideas how to solve it?

A quick win could be to track the direction of the scroll between `scroll.smack` events. Since that event fires each time you scroll, we can check if the `scrollTop` value of the bound element ($this) has increased (scroll down) or decreased (scroll up) since the last event. Something like [this](http://stackoverflow.com/a/4326907/700897).

If we satisfy the conditions for smacking the 'top' edge **and** we're currently scrolling up, then trigger a smack. 

The same should be done for the bottom edge as well. For example, if we smack the bottom threshold (say 80% of the way through) and we don't fetch enough content to push us back up below the 80%, then if we scroll up, we'll accidentally trigger a smack on the bottom edge. In other words, it's the same situation as the top-edge, just applied to the bottom.

> BTW I had problems with installing new tests, I'll try to resolve them and run tests next time.

Let me know what the errors were. You should just be able to do an `npm install` in the root of the project and then open up the `runner.html` file.

---

As an aside, I'm not sure that I follow the use of storing the smack options as data options:  https://github.com/mrjoelkemp/jquery.bum-smack/pull/5/files#diff-419950f0fca14a2b5fc2c6770101e81bR18 and https://github.com/mrjoelkemp/jquery.bum-smack/pull/5/files#diff-419950f0fca14a2b5fc2c6770101e81bR53. Would love to hear your thoughts on it.

**Comment by pawel-piskorz** [comment]
> I was a bit confused by https://github.com/mrjoelkemp/jquery.bum-smack/pull/5/files#diff-419950f0fca14a2b5fc2c6770101e81bR55 as that kills the smack binding if the edge is not 'top' or 'bottom'. It might be better to do an explicit error throw if edge is not 'top', 'bottom', or 'either' (case insensitive).

That's not an error. After an edge is smacked, I set its options to false in bumSmackOptions so it won't be handled any more. When both (top and bottom) bumSmackOptions are set to false, it means that there is no more work left for bum-smack and there is no more need for binding.
How would you code this so it would be easy to understand from reading the code itself?

> A quick win could be to track the direction of the scroll between scroll.smack events.
> [8<]
> The same should be done for the bottom edge as well.

Does it works with current bum-smack? If not, then I suggest we make it another feature, so we could wrap this one up because it's going to be hard to follow it all in one thread.
I think "either" option should also go to a new feature, I have a question about it and I think it shouldn't stop us from finishing this one.

> As an aside, I'm not sure that I follow the use of storing the smack options as data options: https://github.com/mrjoelkemp/jquery.bum-smack/pull/5/files#diff-419950f0fca14a2b5fc2c6770101e81bR18 and https://github.com/mrjoelkemp/jquery.bum-smack/pull/5/files#diff-419950f0fca14a2b5fc2c6770101e81bR53. Would love to hear your thoughts on it.

I didn't wanted to have multiple scroll event listeners for the same element (if you run smack twice, one with edge:top and second with edge:bottom). So I have one, but in this event callback function I needed access to options used when user called smack on the element and using data seemed the simplest one. If you have a better way, please let me know.

Keep in mind that running smack twice with different edge option is not the same as running it once with edge:either. The first version allows you to have different threshold for each edge, while the second one (not yet implemented) uses the same options for both edges.

Thanks for your feedback! :)

**Comment by mrjoelkemp** [comment]
@pawel-piskorz I merged it after adding a few comments for clarification on some of the stuff we were talking about. I understand why you implemented it that way; thanks for the clarification. 

I created a new issue to work on the threshold for 'top' edge binding and skipped those tests. I won't cut a new release until the top threshold is done though. The new minified version is a work in progress.

Thanks for your great work on putting this together! Let me know if you want to tackle the 'top' edge threshold with tests. Otherwise, I'll give it a go.

**Comment by pawel-piskorz** [comment]
Thank you!
I'll try to fix #6.
